src/cr/cube/crunch_cube.py

```diff
@@ -200,82 +200,189 @@ def _as_array(self, include_missing=False, get_non_selected=False,
+
+            if self.mr_dim_ind == (1, 2):
+                # In case of MR wee need to "inflate" mask
+                slice_mask = slice_mask[:, np.newaxis, :, np.newaxis]
+
+            mask[i] = slice_mask
+        res = np.ma.masked_array(res, mask=mask)
+        return res
+
     def _prune_body(self, res, transforms=None):
         '''Prune the result based on margins content.
 
         Pruning is the removal of rows or columns, whose corresponding
         marginal elements are either 0 or not defined (np.nan).
         '''
 
+        if self.ndim > 2:
+            return self._prune_3d_body(res, transforms)
+
         # Note: If the cube contains transforms (H&S), and they happen to
         # have the marginal value 0 (or NaN), they're NOT pruned. This is
         # done to achieve parity with how the front end client works (whaam).
-        row_prune = self._row_prune_indices(transforms)
+        row_margin = self.margin(
+            include_transforms_for_dims=transforms,
+            axis=self.row_direction_axis
+        )
+        if row_margin.ndim > 1:
+            # In case of CAT x MR, we have 2D margin
+            row_margin = np.sum(row_margin, axis=1)
+        row_prune_inds = self._margin_pruned_indices(
+            row_margin,
+            self.inserted_rows_inds if transforms else [],
+        )
 
         if len(self.dimensions) == 1 or len(res.shape) == 1:
             # For 1D, margin is calculated as the row margin.
-            return res[~row_prune]
+            return np.ma.masked_array(res, mask=row_prune_inds)
+            # return res[~row_prune_inds]
 
         # Prune columns by column margin values.
-        col_prune = self._col_prune_indices(transforms)
-        res = res[:, ~col_prune]
+        col_margin = self.margin(
+            include_transforms_for_dims=transforms,
+            axis=self.col_direction_axis,
+        )
+        if col_margin.ndim > 1:
+            # In case of MR x CAT, we have 2D margin
+            col_margin = np.sum(col_margin, axis=0)
+        col_prune_inds = self._margin_pruned_indices(
+            col_margin,
+            self.inserted_col_inds if transforms else [],
+        )
+        mask = self._create_mask(res, row_prune_inds, col_prune_inds)
+        res = np.ma.masked_array(res, mask=mask)
+        return res
 
-        # Prune rows by row margin values.
-        return res[~row_prune, :]
+    @staticmethod
+    def _create_mask(res, row_prune_inds, col_prune_inds):
+        mask_rows = np.repeat(
+            row_prune_inds[:, None], len(col_prune_inds), axis=1
+        )
+        mask_cols = np.repeat(
+            col_prune_inds[None, :], len(row_prune_inds), axis=0
+        )
+        return np.logical_or(mask_rows, mask_cols)
 
     def prune_indices(self, transforms=None):
-        row_indices = self._row_prune_indices(transforms)
+        if len(self.dimensions) >= 3:
+            return self._prune_3d_indices(transforms)
```

Comment:
There looks to be quite a bit of logic that is similar to the next 10ish lines below in `_prune_3d_indices`, I wonder if we could actually always run through the 3d_indices code since a 2d table is identical to one slice of a 3d table. Or am I misunderstanding something about the way that `_prune_3d_indices` is working?